secret-sharing/src/churp/switch.rs

```diff
@@ -0,0 +1,883 @@
+    /// `B(x, me)` or `B(me, y)` during the share reduction or full share
+    /// distribution phase.
+    vv: VerificationVector<D::Group>,
+
+    /// A set of shareholders whose points have been received.
+    shareholders: HashSet<Shareholder>,
+
+    /// A list of encoded shareholders' identities whose points have been
+    /// received.
+    xs: Vec<D::PrimeField>,
+
+    /// A list of received switch points.
+    bijs: Vec<D::PrimeField>,
+}
+
+impl<D> SwitchPoints<D>
+where
+    D: DealerParams,
+{
+    /// Creates a new accumulator for switch points.
+    fn new(
+        threshold: u8,
+        me: &D::PrimeField,
+        vm: VerificationMatrix<D::Group>,
+        kind: DimensionSwitchKind,
+    ) -> Result<Self> {
+        let threshold = threshold as usize;
+        let rows = threshold + 1;
+        let cols = 2 * threshold + 1;
+
+        if vm.dimensions() != (rows, cols) {
+            return Err(Error::VerificationMatrixDimensionMismatch.into());
+        }
+
+        // Precomputing the verification vector speeds up switch point
+        // validation.
+        let (n, vv) = match kind {
+            DimensionSwitchKind::ShareReduction => {
+                let vv = vm.verification_vector_for_x(me);
+                let n = rows;
+                (n, vv)
+            }
+            DimensionSwitchKind::FullShareDistribution => {
+                let vv = vm.verification_vector_for_y(me);
+                let n = cols;
+                (n, vv)
+            }
+        };
+
+        // Wrap the matrix in an option so that we can take it when creating
+        // a player.
+        let vm = Some(vm);
+
+        // We need at least n points to reconstruct the polynomial share.
+        let shareholders = HashSet::with_capacity(n);
+        let xs = Vec::with_capacity(n);
+        let bijs = Vec::with_capacity(n);
+
+        Ok(Self {
+            n,
+            vm,
+            vv,
+            shareholders,
+            xs,
+            bijs,
+        })
+    }
+}
+
+impl<D> SwitchPoints<D>
+where
+    D: DealerParams,
+{
+    /// Checks if a switch point is required from the given shareholder.
+    fn needs_point(&self, id: &Shareholder) -> bool {
+        if self.shareholders.len() >= self.n {
+            return false;
+        }
+        !self.shareholders.contains(id)
+    }
+
+    /// Verifies and adds the given switch point.
+    ///
+    /// Returns true if enough points have been received; otherwise,
+    /// it returns false.
+    fn add_point(&mut self, id: Shareholder, bij: D::PrimeField) -> Result<bool> {
+        if self.shareholders.len() >= self.n {
+            return Err(Error::TooManySwitchPoints.into());
+        }
+        if self.shareholders.contains(&id) {
+            return Err(Error::DuplicateShareholder.into());
+        }
+
+        // The identity of the shareholder doesn't require verification.
+        // If the point is valid, it doesn't matter if it came from a stranger.
```

Comment:
True. Verification is costly, that's for sure.
The current solution is elegant, as it removes the need for the previous committee. If I add this verification, I have to introduce `old_shareholders` and `new_shareholders`. Since remote enclaves will be verified anyway, I will skip this for now and just update the comment. If needed, we can change this latter.

Reply:
Sounds good.